### ISSUE TITLE
[BUG] app::remove_timeout() does not work as expected

### ISSUE THREAD

**Issue by AshfordN** [fltk-rs/fltk-rs#1088]
## Describe the bug
Calling `app::remove_timeout()` with a closure that was previously passed to `app::add_timeout()` does not stop the timeout from firing; or the closure from executing.

## To Reproduce
```rust
use fltk::{prelude::*, *};

fn main() {
    let root = app::App::default();

    let mut win = window::Window::default()
        .with_size(400, 300)
        .center_screen();

    win.end();
    win.show();

    app::add_timeout(1.0, callback);
    app::remove_timeout(callback);

    root.run().unwrap();
}

fn callback() {
    println!("fired");
}
```

## Expected behavior
Calling `app::remove_timeout()` with a closure that was previously passed to `app::add_timeout()` should prevent the timeout from firing, thereby preventing the closure's execution.

## Screenshots
NA

## Desktop info
 - OS: Linux Mint 20.2 Cinnamon 64bit
 - Version: 1.2.25

## Additional info
I initially tried this with Box'ed closures and had the same issue, but the example above demonstrates the issue in its simplest and most unmistakable form.
From my understanding, `remove_timeout()`, in the underlying FLTK API, works by searching the list of timeouts to find one (or more) that is associated with the same function pointer and data pointer that you're passing in and then removing it from the list. However, the issue here is twofold:

1.  `app::add_timeout()` and `app::remove_timeout()` each define local shim functions that naturally have different address locations across the two functions. This could be solved by defining a common shim for both functions, in the global scope.
2. The `data` variables constructed locally, in `app::add_timeout()` and `app::remove_timeout()`, end up pointing to different memory locations, even for successive calls to the same function. So far, I haven't quite figured out how to solve this.

I came up with [this patch](https://github.com/AshfordN/fltk-rs/commit/d788e6b71c6ba3356ff58c8079f07c6c699c857b) as a potential solution -- that works on the basis of a returned timeout handle -- but I'm not sure if this is the best solution. 
As usual, I'd like your input, and if you're satisfied with the fix, I'll open a PR.

## Build failures
NA

**Comment by MoAlyousef** [comment]
Hi
So it's a problem that Rust's closures aren't comparable. For simple function pointers, there's the add_timeout2 and remove_timeout2: 
```rust
    fn callback() {
        println!("fired");
    }
    app::add_timeout2(1.0, callback);
    app::remove_timeout2(callback);
```
An important use case for timeouts is to be able to use them in the callback function:
```rust
    fn callback() {
        println!("fired");
        if true {
            app::repeat_timeout2(1.0, callback);
        } else {
            app::remove_timeout2(callback);
        }
    }
```

For closures, your changes address the first issue, but not the second. While add_timeout/remove_timeout address the second but not the first:
```rust
fn move_button(mut btn: button::Button) {
    btn.set_size(btn.w() - 2, btn.h());
    btn.parent().unwrap().redraw();
    if btn.w() == 0 {
        app::remove_timeout(move || {
            let btn = btn.clone();
            move_button(btn);
        });
    } else {
        app::repeat_timeout(0.3, move || {
            let btn = btn.clone();
            move_button(btn);
        });
    }
}

btn.set_callback(|b| {
    let btn = b.clone();
    app::add_timeout(0.016, move || {
        let btn = btn.clone();
        move_button(btn)
    });
    let btn = b.clone();
    app::remove_timeout(move || {
        let btn = btn.clone();
        move_button(btn);
    });
});
```

I've modified add_timeout3 where the handle is passed as an argument so both cases can be addressed, but I feel it's a bit clunky:
```rust
    fn move_button(mut btn: button::Button, handle: *mut Box<dyn FnMut()>) {
        btn.set_size(btn.w() - 2, btn.h());
        btn.parent().unwrap().redraw();
        if btn.w() == 20 {
            app::remove_timeout3(handle);
        } else {
            app::repeat_timeout(0.3, move || {
                let btn = btn.clone();
                move_button(btn, handle);
            });
        }
    }

    let handle = app::add_timeout3(1.0, move |handle| {
        let btn = btn.clone();
        move_button(btn, handle)
    });
    app::remove_timeout3(handle);
```

The modifications:
```patch
diff --git a/fltk/src/app/rt.rs b/fltk/src/app/rt.rs
index 1771063c..1c3499de 100644
--- a/fltk/src/app/rt.rs
+++ b/fltk/src/app/rt.rs
@@ -275,15 +275,17 @@ pub fn has_idle2(cb: fn()) -> bool {
 }
 
 unsafe extern "C" fn timeout_shim(data: *mut raw::c_void) {
-    let a: *mut Box<dyn FnMut()> = data as *mut Box<dyn FnMut()>;
-    let f: &mut (dyn FnMut()) = &mut **a;
-    let _ = panic::catch_unwind(panic::AssertUnwindSafe(f));
+    let a: *mut Box<dyn FnMut(*mut Box<dyn FnMut()>)> = data as *mut Box<dyn FnMut(*mut Box<dyn FnMut()>)>;
+    let f: &mut (dyn FnMut(*mut Box<dyn FnMut()>)) = &mut **a;
+    let _ = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        (*f)(data as _)
+    }));
 }
 
-pub fn add_timeout3<F: FnMut() + 'static>(tm: f64, cb: F) -> *mut Box<dyn FnMut()> {
+pub fn add_timeout3<F: FnMut(*mut Box<dyn FnMut()>) + 'static>(tm: f64, cb: F) -> *mut Box<dyn FnMut()> {
     unsafe {
         assert!(crate::app::is_ui_thread());
-        let a: *mut Box<dyn FnMut()> = Box::into_raw(Box::new(Box::new(cb)));
+        let a: *mut Box<dyn FnMut(*mut Box<dyn FnMut()>)> = Box::into_raw(Box::new(Box::new(cb)));
         let data: *mut raw::c_void = a as *mut raw::c_void;
         let callback: Option<unsafe extern "C" fn(arg1: *mut raw::c_void)> = Some(timeout_shim);
         fl::Fl_add_timeout(tm, callback, data);
@@ -294,7 +296,7 @@ pub fn add_timeout3<F: FnMut() + 'static>(tm: f64, cb: F) -> *mut Box<dyn FnMut(
             timeout_shim as *const (), data
         );
 
-        a
+        data as _
     }
 }

````

**Comment by AshfordN** [comment]
> I've modified add_timeout3 where the handle is passed as an argument so both cases can be addressed
I've applied your patch and added `app::repeat_timeout3()` as a new companion function.

> The clinkiness can probably be hidden by wrapping the closure arg in an opaque type TimeoutHandle

I don't quite understand what you mean here. Are you referring to a type alias like:
```rust
type TimeoutHandle = *mut Box<dyn FnMut()>;
```
or something different?

Also, looking at the other functions defined in that file, I think there might be more function pairs affected by similar issues; such as `app::add_idle()` and `app::remove_idle()`.

**Comment by MoAlyousef** [comment]
Yes an alias would work. And you're right, add_idle and remove_idle suffer from the same problem.

**Comment by AshfordN** [comment]
Another thing though, isn't the use of `Box::into_raw()` in these functions leaking memory?

**Comment by MoAlyousef** [comment]
Basically the closure is turned into a fat pointer, leaked as a void pointer into a static shim where it's cast back into a boxed function which is called. The lifetime of the function is expected to be static. Now with widgets which actually subclass FLTK widgets, we're able to store these void pointer representing data and free them with the widget's destructor. The Fl object is a static object initialized by FLTK before main, so we can't subclass it, i.e. we can't store these void pointers and we can't delete them. 

Now with the remove_timeout3, since we have the handle, it can potentially be freed (by calling Box::from_raw on the passed handle). 

In any case, it doesn't cause an increasing memory usage with time, and valgrind doesn't seem to consider it a leak. It detects other leaks mostly internal to how pango deals with fonts.

**Comment by AshfordN** [comment]
Fixed by PR #1090
